sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs

```diff
@@ -14,10 +14,18 @@ namespace Azure.Messaging.EventHubs.Producer
     ///   intended to be sent to the Event Hubs service as a single batch.
     /// </summary>
     ///
+    /// <remarks>
+    ///   The operations for this class are thread-safe and will prevent changes when
+    ///   actively being published to the Event Hubs service.
+    /// </remarks>
+    ///
     public sealed class EventDataBatch : IDisposable
     {
+        /// <summary>An object instance to use as the synchronization root for ensuring the thread-safety of operations.</summary>
+        private readonly object SyncGuard = new object();
```

Comment:
Just double checking, but should we be concerned about `AsEnumerable<T>` not in sync? I'm assuming we'll lock the batch before calling it.

Reply:
Yeah, we lock the batch and only then call for the enumerable, which is an internal-only use.   I don't want to hold the sync lock during the entire send operation nor have to acquire it multiple times in the case of retries.   I'm comfortable with the trade-off in this case.

If we wanted to be absolutely sure and bullet-proof in the face of some future change that may forget to lock the batch, we could acquire the lock here.... but if something goes wrong, the enumerator will throw because "the collection has been modified" and the operation will fail.   I'm going to leave this as-is.